Confer outputs arity implicitly for list/singular

werkt opened the pull request:
Multiple outputs, expanded through placeholder specification of list
attributes, should be conveyed to outputs struct definition unmodified.
Presumed singleton extraction from implicit outputs subsystem has been
reworked, and tests of singular/list behavior have been updated.

werkt commented:
The previous pull request was closed due to a conflict with internal API dependencies, and RuleContext was singled out as an interface conflict, but it made *no* interface modifications (no signature changes for any methods).

If there is an internal conflict for one of the other interface changes that occur in other classes, please identify them, and I'd be happy to split out the functionality for new usage.

ulfjack commented:
Assigning to laurentlb, since this is intended to provide a Skylark API.

dslomov commented:
Please explain what changed in Skylark API. Skylark design changes should go through https://bazel.build/designs/skylark/skylark-design-process.html

mhlopko commented:
Friendly ping, is this still under development? @laurentlb does this change look good to you, is only the process that needs to be done?

werkt commented:
I still really want this to go in. I've been otherwise occupied behind some
other projects, but my next action item is to implement this as a specific
function as was suggested in the design document

On Oct 11, 2017 8:20 AM, "Marcel Hlopko" <notifications@github.com> wrote:

> Friendly ping, is this still under development? @laurentlb
> <https://github.com/laurentlb> does this change look good to you, is only
> the process that needs to be done?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/bazel/pull/3448#issuecomment-335791179>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABTluzu71AZhhjXK35NHmMMkmUIm5f8ks5srLJ7gaJpZM4Oix7q>
> .
>

buchgr commented:
Ping @laurentlb :)

laszlocsomor commented:
ping -- are there any updates? can we merge or close this PR?

werkt commented:
I've updated the change. The discussion fizzled in the document, and the request was made to move this to a functional behavior inside of context, which is actually a net product of this change already, I just specifically wanted the implicit outputs behavior so that it was more difficult to use incorrectly.

Personally, I say that the implicit outputs behavior has not been deprecated, removed, or even altered in the year since this was proposed, and all this does is shore up a case that currently where bazel has undefined behavior. If the team doesn't want to accept it, I accept that, but I think it does no more harm than good (and makes my rule definitions a little cleaner). I will get the CI to pass and leave it to the team to decided as it stands.

laszlocsomor commented:
Thanks for the update, @werkt !

@laurentlb , @dslomov : ping

dslomov commented:
Output maps are full of madness: https://docs.google.com/document/d/1ic9lJPn-0VqgKcqSbclVWwYDW2eiV-9k6ZUK_xE6H5E/
We really want to deprecate and remove them in favor of output groups.
So let's not add more fuel to this fire.  Sorry.
